### PR TITLE
fixed and refactored balance effect

### LINKED ISSUE #6593: [[Balance]] not functioning properly
Balance is properly sacrificing lands and discarding cards, but not sacrificing creatures. Opponent has more than zero creatures out while I have zero and it is not forcing him to sacrifice his creatures when I cast this card.

### PR DESCRIPTION
**Pull request by 18ths**
fixed #6593 - the problem was obvious copy-paste problem - instead of FILTER_CONTROLLED_CREATURE, FILTER_CONTROLLED_PERMANENT_LAND was used. fixed the issue, and removed duplicated code so it wont appear in future. Also, now it's unit tested.

### CODE REVIEW

**Comment by Zerrisx** [comment]
Does this happen to also fix this issue from #6532?
image

**Comment by 18ths** [comment]
I think so. We we're calculating the land count (in your case 0) and applying it on creatures.

**Comment by 18ths** [comment]
@jeffwadsworth - I've noticed you self-assigned #6532. This fixes one of issues listed there (as well as #6593), could you review this PR, so our work wont be duplicated? :)

File: Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java [hunk 1]
```diff
@@ -108,34 +86,41 @@ public boolean apply(Game game, Ability source) {
...
-        //Cards in hand
-        int minCard = Integer.MAX_VALUE;
-        Map<UUID, Cards> cardsToDiscard = new HashMap<>(2);
+        return true;
+    }
+
+    private Cards getPermanentsToSacrifice(Game game, Ability source, Player controller,
+                                           FilterControlledPermanent filterPermanent, FilterControlledPermanent filterPermanentDialog) {
+        int min = Integer.MAX_VALUE;
+        Cards permToSacrifice = new CardsImpl();
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java:115]: `Cards` is hashset, so sacrifice order and logs must be random. The best way -- use simple list -- then logs will be "sorted" by choosed players (more user friendly).

Hmm. There are oracle text: Each player will know the choices made by the players who chose before them. So you must:
1. Collect chooses to keep/sacrifice lands (inform other players about chooses after each cycle);
2. Sacrifice;
3. Repeat 1-2 for creatures;
4. Repeat 1-2 for hands.

Full oracle text:

First the player whose turn it is chooses which lands (if any) to keep, then each other player in turn order does the same. Each player will know the choices made by the players who chose before them. All of the unchosen lands are then sacrificed simultaneously. Then the process is repeated for cards in hand, except that no cards are revealed until all players have chosen what to discard, at which point those cards are all discarded simultaneously. Lastly, the process is repeated for creatures, and players will again know earlier choices made when deciding what to sacrifice. All of the unchosen creatures are then sacrificed simultaneously. (2016-06-08)

File: Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java [hunk 2]
```diff
@@ -144,25 +129,17 @@ public boolean apply(Game game, Ability source) {
             if (player == null) {
                 continue;
             }
-            Cards cards = new CardsImpl();
-            TargetCardInHand target = new TargetCardInHand(minCard, filter3);
-            if (!target.choose(Outcome.Discard, player.getId(), source.getSourceId(), game)) {
+            TargetControlledPermanent target = new TargetControlledPermanent(min, min, filterPermanentDialog, true);
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java:132]: Wrong logic. `Min` is card's amount to keep, not sacrifice. So you must use like `sacCount = countAll() - min`. Choose to sac.

BUT card's oracle text talk about choose to keep. So I recommends to change choose logic to it -- players must choose cards to keep (with good outcome for AI), not to sacrifice.

**Comment by 18ths** [comment]
@jeffwadsworth @JayDi85 It should be okay now. I inversed the logic as suggested and added information to players after every sacrifice batches. One thing I'm not sure: I've changed the outcomes for AI, because the previous ones would make much sense now that the logic is inversed. (If AI is asked to pick the cards, it should be able to protect its most worth cards in this moment instead of get advantage of sacrifice/discard effects) Do you think Outcome.Protect makes sense in that context?

**Comment by JayDi85** [comment]
You must look at good or bad. If it have name then use it (protect is ok). If it haven't then use Detriment (bad) or Benefit (good). So "keep card" is good action.

Bad or good equals "is target/choice will be good or bad for targeted permanent/card". So AI choose targets by priority:
* if outcome is good then it selects own cards first (from big to small) then opponents cards (from small to big);
* if outcome is bad then it selects opponent cards first (from big to small) then own cards (from small to big).

**Pull request merged**